### DISCUSSION TITLE
Validating serverVerificationData  (token) from flutter

### DISCUSSION THREAD

**Discussion by MerNat** [awa/go-iap#234] [Q&A] [answered]
I'm having this token from flutter ([in_app_purchase](https://pub.dev/packages/in_app_purchase)) PurchasDetails response.

I do have those three fields: 

```
productID, verificationData.source & verficationData.serverVerficationData fields.
```
It's known that I have to send the serverVerificationData to the backend for verification but this package is not clear enough to verify a purchase based on this data (token) from the client side. 

Or probably I might have not understood the functions overall. 

How do I verify this data using this package/library?

**Comment by richzw** [comment]
First of all, this package is used for server verification. 

If you want to verify the transaction receipt on Flutter client side, you could try 

```
     Future<http.Response> validateReceiptIos(receiptBody,isTest) async {
          final String url = isTest
            ? 'https://sandbox.itunes.apple.com/verifyReceipt'
             : 'https://buy.itunes.apple.com/verifyReceipt';
           return  await http.post(
           Uri.parse(url),
            headers: {
            'Accept': 'application/json',
             'Content-Type': 'application/json',
              },
            body: json.encode(receiptBody),
              );
             }
```

refer to https://stackoverflow.com/questions/59617454/flutter-how-to-verify-inapppurchases-ios

I am not good at the Flutter, Please correct me if something is wrong or missing.

**Reply by MerNat** [reply]
Thanks for the reply @richzw 
Though verification is handled on the backend on my side. Thus I'm using this library to see if I can verify the the transaction from the backend side. Also, this endpoint from Apple is deprecated. (Link: https://developer.apple.com/documentation/appstorereceipts/verifyreceipt)

**Reply by MerNat** [reply]
This is a package library from dart but thinking if it might exist in this library also.

https://github.com/flutter/codelabs/blob/main/in_app_purchases/step_09/dart-backend/lib/app_store_purchase_handler.dart#L48

**Reply by richzw** [reply]
Yes, you could use the `get transaction info` of [app server API](https://developer.apple.com/documentation/appstoreserverapi) to verify the transaction by validating the transaction ID, here is one sample

```
	transactionId := "FAKETRANSACTIONID"
	a := api.NewStoreClient(c)
	ctx := context.Background()
	response, err := a.GetTransactionInfo(ctx, transactionId)

	transantion, err := a.ParseSignedTransaction(response.SignedTransactionInfo)
	if err != nil {
	    // error handling
	}

	if transaction.TransactionId == transactionId {
		// the transaction is valid
	}
```

**Reply by MerNat** [reply]
But the issue is I'm not getting the `transactionId` from the client side (flutter ~ in_app_purchase package).
it seems they're also verifying just like the same one you shared.
https://codelabs.developers.google.com/codelabs/flutter-in-app-purchases#9

**Reply by richzw** [reply, marked as answer]
Sorry for that I am not familiar with flutter, maybe this answer https://stackoverflow.com/questions/73035526/how-to-get-a-unique-identifier-for-in-app-purchases-in-flutter-which-stays-alway could help you to retrieve transactionId from receipt in flutter.

**Reply by MerNat** [reply]
That's a huge help. Thanks @richzw
